climpred/smoothing.py

```diff
@@ -0,0 +1,157 @@
+import xarray as xr
+
+try:
+    import xesmf as xe
+
+    def spatial_smoothing_xesmf(ds, boxsize=(5, 5)):
+        """Apply spatial smoothing by regridding to `boxsize` grid.
+
+        Reference:
+          * Goddard, L., A. Kumar, A. Solomon, D. Smith, G. Boer, P.
+                Gonzalez, V. Kharin, et al. “A Verification Framework for
+                Interannual-to-Decadal Predictions Experiments.” Climate
+                Dynamics 40, no. 1–2 (January 1, 2013): 245–72.
+                https://doi.org/10/f4jjvf.
+
+        Args:
+            ds (xr.object): input. xr.DataArray prefered.
+            boxsize (tuple): target grid.
+                             Defaults to `(5, 5)` (see Goddard et al. 2013).
+
+        Returns:
+            ds_smoothed (xr.object): boxsize-regridded input
+
+        """
+        if isinstance(ds, xr.Dataset):
+            was_dataset = True
+            ds = ds.to_array().squeeze()
+        else:
+            was_dataset = False
+        ds_out = xe.util.grid_global(*boxsize)
+        regridder = xe.Regridder(ds, ds_out, 'bilinear')
```

Comment:
I would suggest a keyword for remap type. This could be 'bilinear', 'patch', etc. It will be hard to natively support conservative remapping since you need grid cell vertices for that.

Reply:
Also, grid_global?

Maybe you can adapt my code snippet
```
def _regrid_it(da, d_lon, d_lat, **kwargs):
    '''
    Global 2D rectilinear grid centers and bounds
     Parameters
    ----------
    da : xarray DataArray
        Contain input and output grid coordinates. Look for variables
        ``lon``, ``lat``, and optionally ``lon_b``, ``lat_b`` for
        conservative method.
         Shape can be 1D (Nlon,) and (Nlat,) for rectilinear grids,
        or 2D (Ny, Nx) for general curvilinear grids.
        Shape of bounds should be (N+1,) or (Ny+1, Nx+1).
     d_lon : float
        Longitude step size, i.e. grid resolution
     d_lat : float
        Latitude step size, i.e. grid resolution
     Returns
    -------
    da : xarray DataArray with coordinate values
     '''
    if not warn_lon_lat_dne(da):  # both lon and lat exist
        grid_out = {LON: np.arange(da[LON].min(), da[LON].max() + d_lon, d_lon),
                    LAT: np.arange(da[LAT].min(), da[LAT].max() + d_lat, d_lat)}
        regridder = xe.Regridder(da, grid_out, **kwargs)
        return regridder(da)
    else:
        return da


 def regrid_it(ds, d_lon=1, d_lat=None, method='bilinear',
              periodic=False, filename=None, reuse_weights=True):
    '''
    Quick regridding
     Parameters
    ----------
    ds : xarray DataSet
        Contain input and output grid coordinates. Look for variables
        ``lon``, ``lat``, and optionally ``lon_b``, ``lat_b`` for
        conservative method.
         Shape can be 1D (Nlon,) and (Nlat,) for rectilinear grids,
        or 2D (Ny, Nx) for general curvilinear grids.
        Shape of bounds should be (N+1,) or (Ny+1, Nx+1).
     d_lon : float, optional
        Longitude step size, i.e. grid resolution; if not provided,
        will equal 1
     d_lat : float, optional
        Latitude step size, i.e. grid resolution; if not provided,
        will equal d_lon
     method : str
        Regridding method. Options are
         - 'bilinear'
        - 'conservative', **need grid corner information**
        - 'patch'
        - 'nearest_s2d'
        - 'nearest_d2s'
     periodic : bool, optional
        Periodic in longitude? Default to False.
        Only useful for global grids with non-conservative regridding.
        Will be forced to False for conservative regridding.
     filename : str, optional
        Name for the weight file. The default naming scheme is::
             {method}_{Ny_in}x{Nx_in}_{Ny_out}x{Nx_out}.nc
         e.g. bilinear_400x600_300x400.nc
     reuse_weights : bool, optional
        Whether to read existing weight file to save computing time.
        False by default (i.e. re-compute, not reuse).
     Returns
    -------
    ds : xarray DataSet with coordinate values or DataArray
     '''
    if d_lat is None:
        d_lat = d_lon

     kwargs = {
        'd_lon': d_lon,
        'd_lat': d_lat,
        'method': method,
        'periodic':periodic,
        'filename': filename,
        'reuse_weights': reuse_weights,
    }

     if isinstance(ds, xr.Dataset):
        ds = xr.merge(_regrid_it(ds[var], **kwargs)
                      for var in ds.data_vars)
    else:
        ds = _regrid_it(ds, **kwargs)

     return ds
```

Reply:
I am unsure whether we need to have this long function with documentation in our codebase. couldnt we just import this: `from xesmf.utils import regrid_it`. 1x1 deg is default here which is also ok for me. and then for the goddard wrapper I would adapt for 5x5. the difference is then that here it doesnt use dicts but `d_lon/lat` keywords with stepsize. `temporal_smoothing` can also be adapted for this. its either `time` or `init`. we dont allow for anything else and this the function can guess. any objections for this implementation?

Reply:
We can't import it because it doesn't exist in the codebase (the PR was rejected)

Reply:
ah ok. so is this the only way to have not automatically remapping to global grid?